### PR TITLE
Adding EMCF solver

### PR DESCRIPTION
**Pull request by piyushrpt**
* Unwrapping in time first and then in space
* Support for computing residues and flood filling using gradients
* Support for including models when computing residues
* Example usage can be found in `tests/workflows/test_emcf.py`

### CODE REVIEW

File: src/spurt/workflows/emcf/_solver.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+        wrap_data: np.ndarray
+            2D array of size (nslc, npoints) or (nifg, npoints).
+
+        Returns
+        -------
+        uw_data: np.ndarray
+            2D float32 array of size (nifg, npoints).
+        """
+        if wrap_data.ndim != 2:
+            errmsg = f"Input data has more than two dimension - {wrap_data.ndim}"
```
Comment (mattcalef) [inline src/spurt/workflows/emcf/_solver.py:96]: [NB] It could be that `wrap_data.ndim == 1`, in which case the error message will not make sense.

File: src/spurt/workflows/emcf/_solver.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+        else:
+            errmsg = f"Unknown cost type: {self.settings.t_cost_type}"
+            raise ValueError(errmsg)
+
+        # Create output array
+        grad_space = np.zeros((self.nifgs, self.nlinks), dtype=np.float32)
+
+        print(f"Temporal: Number of interferograms: {self.nifgs}")
+        print(f"Temporal: Number of links: {self.nlinks}")
+        print(f"Temporal: Number of cycles: {self._solver_time.ncycles}")
```
Comment (mattcalef) [inline src/spurt/workflows/emcf/_solver.py:145]: At some point we might consider logging.

File: src/spurt/workflows/emcf/_settings.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    worker_count: int = 0
+    points_per_batch: int = 1000
+    t_cost_type: str = "unit"
+    t_cost_scale: float = 100.0
+    s_cost_type: str = "unit"
+    s_cost_scale: float = 100.0
+
+    def __post_init__(self):
+        assert self.t_cost_type in ["unit", "distance", "centroid"]
+        assert self.s_cost_type in ["unit", "distance", "centroid"]
```
Comment (mattcalef) [inline src/spurt/workflows/emcf/_settings.py:38]: [NB] Perhaps we might want to assert `points_per_batch` is positive?

Similarly may want to assert `s_cost_scale` and `t_cost_scale` are not going to cause errors.

**Review by mattcalef — APPROVED** [review]
This looks good. I've added a few non-blocking issues, which we can tackle later.

File: src/spurt/workflows/emcf/_solver.py (hunk 1 above)
Comment (piyushrpt) [reply src/spurt/workflows/emcf/_solver.py:96]: Done

File: src/spurt/workflows/emcf/_solver.py (hunk 2 above)
Comment (piyushrpt) [reply src/spurt/workflows/emcf/_solver.py:145]: I found this: https://github.com/isce-framework/snaphu-py/issues/15
Not sure if there is another mechanism they would like us to use yet.

File: src/spurt/workflows/emcf/_settings.py (hunk 3 above)
Comment (piyushrpt) [reply src/spurt/workflows/emcf/_settings.py:38]: Done

**Comment by piyushrpt** [comment]
I will leave this branch till @gmgunter gets a chance to take a better look. No rush. The feature branch should be fully functional.

File: src/spurt/workflows/emcf/_solver.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+            )
+
+            # Update the spatial gradients with estimated flows
+            grad_space[:, i_start:i_end] += 2 * np.pi * flows.T
+
+        return grad_space
+
+    def unwrap_gradients_in_space(self, grad_space: np.ndarray) -> np.ndarray:
+        """Spatially unwrap each interferogram sequentially."""
+        # First set up temporal cost
```
Comment (gmgunter) [inline src/spurt/workflows/emcf/_solver.py:225]:
```suggestion
        # First set up spatial cost
```

File: src/spurt/workflows/__init__.py [hunk 5]
```diff
@@ -0,0 +1 @@
+from . import emcf
```
Comment (gmgunter) [inline src/spurt/workflows/__init__.py:1]:
```suggestion
from . import emcf

__all__ = [
    "emcf",
]
```

File: src/spurt/workflows/emcf/_solver.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+        elif self.settings.t_cost_type == "centroid":
+            cost = utils.centroid_costs(
+                self._solver_time.points,
+                self._solver_time.cycles,
+                self._solver_time.dual_edges,
+                scale=self.settings.t_cost_scale,
+            )
+        else:
+            errmsg = f"Unknown cost type: {self.settings.t_cost_type}"
+            raise ValueError(errmsg)
```
Comment (gmgunter) [inline src/spurt/workflows/emcf/_solver.py:138]: None of these temporal unwrapping cost functions match the actual cost function that's used in the cited paper. That's fine, but this discrepancy between the algorithm described in the paper and what's actually implemented here should be mentioned in the docstring so that we don't surprise users expecting a faithful implementation of the EMCF algorithm.

Do you have any plans to implement Pepe & Lanari's temporal unwrapping cost model as well in the future?

File: src/spurt/links/_interface.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        """Return number of observations in model."""
+
+    @property
+    def ndim(self) -> int:
+        """Return number of model parameters."""
+
+    def estimate_model(
+        self, wrapdata: np.ndarray, weights: np.ndarray | None = None
+    ) -> tuple[float | np.ndarray, np.ndarray, float]:
+        """Return estimated model parameters, reconstructed model and quality metric."""
```
Comment (gmgunter) [inline src/spurt/links/_interface.py:46]: Can you add a `Returns` section with a more detailed description of each of these outputs? Right now, I'm not sure how to interpret this stuff.

What are the model parameters? What does the quality metric signify? Should the quality metric be between 0 and 1? What are the units of the reconstructed model? Phase in radians? Or displacement in meters? This is the kind of stuff that should be made more explicit.

File: src/spurt/mcf/_interface.py [hunk 8]
```diff
@@ -96,8 +165,28 @@ def compute_residues(self, wrapdata: ArrayLike) -> ArrayLike:
...
+            graph. Includes the grounding node.
+        """
+
     def residues_to_flows_many(
-        self, residues: ArrayLike, cost: ArrayLike, revcost: ArrayLike | None = None
+        self,
+        residues: ArrayLike,
+        cost: ArrayLike,
+        revcost: ArrayLike | None = None,
+        worker_count: int | None = None,
```
Comment (gmgunter) [inline src/spurt/mcf/_interface.py:189]: Let's add a description for the `worker_count` argument to the docstring below

**Review by gmgunter — COMMENTED** [review]
Thanks for your patience! I'm starting to catch back up now

**Comment by piyushrpt** [comment]
I think the comments here have come in a little too late to address them satisfactorily and not impact the two other PRs that are queued up. I can included them in the last of the PRs that are stringed together. Does that work?

File: src/spurt/workflows/emcf/_solver.py (hunk 6 above)
Comment (piyushrpt) [reply src/spurt/workflows/emcf/_solver.py:138]: None of their code is in the public domain, so naturally it is not going to be possible to replicate it as is. They also use constrained Delaunay triangulation etc - so I would say noone should expect a faithful implementation, since none of it is available in the public domain. The framework allows for coherence related costs to be applied - its a question to this team, if we are guaranteed access to these as inputs in all cases. If so, we can consider adding that as an option.

**Comment by gmgunter** [comment]
>Does that work?

Yeah, that's totally fine with me. I'll continue adding comments as I read through the changes, but they can be addressed in subsequent PRs. I'll approve each of the backlog PRs once I'm finished reviewing it, even if there are remaining comments, until I'm caught back up with you.

**Review by gmgunter — APPROVED** [review]
Thanks, LGTM, let's go ahead and merge this one. I appreciate it if you could address the comments below in a subsequent PR (or respond to them if they're off-base).

File: src/spurt/mcf/_ortools.py [hunk 9]
```diff
@@ -93,29 +115,59 @@ def _prepare_dual(self) -> None:
...
-        # Set supply of ground_node
+        residues = np.rint(residues / (2 * np.pi)).astype(int)
+        residues[0] = -np.sum(residues[1:])
+        return residues
+
+    def compute_residues_from_gradients(
+        self,
+        graddata: ArrayLike,
+    ) -> ArrayLike:
+        """Compute phase residues for one set of real input gradients."""
```
[the remark this answers is not in the document]
Comment (piyushrpt) [reply src/spurt/mcf/_ortools.py:150]: Done

File: src/spurt/mcf/_ortools.py [hunk 10]
```diff
@@ -93,29 +115,59 @@ def _prepare_dual(self) -> None:
...
+        if graddata.size != self.nedges:
+            errmsg = (
+                f"Size mismatch for residue computation."
+                f" Received {graddata.shape} with {self.nedges} edges"
+            )
+            raise ValueError(errmsg)
+
+        cyc0 = np.abs(self.dual_edges[:, 0])
+        cyc1 = np.abs(self.dual_edges[:, 1])
```
[the remark this answers is not in the document]
Comment (piyushrpt) [reply src/spurt/mcf/_ortools.py:159]: Done

File: src/spurt/mcf/_ortools.py [hunk 11]
```diff
@@ -93,29 +115,59 @@ def _prepare_dual(self) -> None:
...
+        cyc0_dir = self.dual_edge_dir[:, 0]
+        cyc1_dir = self.dual_edge_dir[:, 1]
+        grad_sum = np.zeros(self.ncycles + 1, dtype=np.float32)
+        # add.at to handle repeated indices
+        np.add.at(grad_sum, cyc0, cyc0_dir * graddata)
+        np.add.at(grad_sum, cyc1, cyc1_dir * graddata)
+
+        residues = np.rint(grad_sum / (2 * np.pi))
```
[the remark this answers is not in the document]
Comment (piyushrpt) [reply src/spurt/mcf/_ortools.py:167]: Done

File: src/spurt/mcf/utils.py [hunk 12]
```diff
@@ -62,16 +68,29 @@ def flood_fill(indata: np.ndarray, links: np.ndarray, flows: np.ndarray):
         errmsg = f"Dimension mismatch - {links.shape} vs {flows.shape}"
         raise ValueError(errmsg)
 
+    if len(indata) == len(links):
```
[the remark this answers is not in the document]
Comment (piyushrpt) [reply src/spurt/mcf/utils.py:71]: This is a highly unlikely edge case. Any suggestions for how to handle this? Add another kwarg that by default is set to "auto" or else explicitly set to "points" or "links"?

File: test/mcf/test_ortools_mcf.py [hunk 13]
```diff
@@ -163,3 +179,42 @@ def test_unwrap_many_oneworker():
...
+    for ii, edge in enumerate(solver.edges):
+        edge_to_index[spurt.graph.utils.order_points((edge[0], edge[1]))] = ii
+
+    cyc = solver.cycles[0]
+
+    for ii in range(3):
+        jj = (ii + 1) % 3
+        ind = edge_to_index[spurt.graph.utils.order_points((cyc[ii], cyc[jj]))]
+        print(ind, grads[ind], solver.dual_edges[ind], solver.dual_edge_dir[ind])
+    """
```
[the remark this answers is not in the document]
Comment (piyushrpt) [reply test/mcf/test_ortools_mcf.py:206]: Happy to take them out. Usually, these are handy when you want to debug something in detail when things go wrong. There doesn't seem to be another location in the repo to include utilities like this.

[the remark this answers is not in the document]
Comment (piyushrpt) [reply test/mcf/test_ortools_mcf.py:206]: Done

File: src/spurt/mcf/utils.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (gmgunter) [reply src/spurt/mcf/utils.py:71]: Yeah that sounds like a good solution to me. I think we should prefer to avoid using the "auto" mode, though, in case this issue ever comes up in practice.

File: test/mcf/test_ortools_mcf.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (gmgunter) [reply test/mcf/test_ortools_mcf.py:206]: >Usually, these are handy when you want to debug something in detail when things go wrong.

Sorry I didn't read through the code carefully since it was commented out. I kind of think its fine to just leave print statements like this uncommented in tests. By default, `pytest` suppresses the output of tests that pass. So these print statements won't affect test output verbosity in the typical case where everything is working fine. You'll only see them when the test fails, which is exactly when you'd want them.

Anyway, either approach is fine with me (removing them or uncommenting them). But keeping the debugging code commented out wouldn't be my preference.

File: src/spurt/mcf/utils.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (piyushrpt) [reply src/spurt/mcf/utils.py:71]: Have added this to `feat/dirichlet`
